### PR TITLE
fcitx5: substitute path in all desktop files and dbus service files

### PR DESCRIPTION
**Pull request by duament**
## Description of changes

Fcitx5 does not load addons properly if activated by dbus.

This PR substitutes the Exec path in fcitx5 dbus service file to fix it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by duament** [comment]
> Not sure why the desktop file uses an absolute path (see #308324) in the first place. As for the dbus service I'm admittedly not experienced enough with that to comment.

Not sure why. But these absolute paths are set by the upstream. 

https://github.com/fcitx/fcitx5/blob/5.1.9/data/org.fcitx.Fcitx5.desktop.in.in

https://github.com/fcitx/fcitx5/blob/5.1.9/src/modules/dbus/org.fcitx.Fcitx5.service.in

File: pkgs/tools/inputmethods/fcitx5/with-addons.nix [hunk 1]
```diff
@@ -38,10 +38,13 @@ symlinkJoin {
...
+    dbus=share/dbus-1/services/org.fcitx.Fcitx5.service
     rm $out/$desktop
     rm $out/$autostart
+    rm $out/$dbus
     cp ${fcitx5}/$desktop $out/$desktop
-    sed -i $out/$desktop -e "s|^Exec=.*|Exec=$out/bin/fcitx5|g"
+    cp ${fcitx5}/$dbus $out/$dbus
+    substituteInPlace $out/$desktop $out/$dbus --replace-fail ${fcitx5} $out
     ln -s $out/$desktop $out/$autostart
   '';
```
Comment (Vonfry) [inline pkgs/tools/inputmethods/fcitx5/with-addons.nix:49]: Would you like to update other desktop files to addons one as well, though they are the same thing now?

Comment (duament) [reply pkgs/tools/inputmethods/fcitx5/with-addons.nix:47]: Absolutely. Updated.

**Review by Vonfry — APPROVED** [review]

File: pkgs/tools/inputmethods/fcitx5/with-addons.nix [hunk 2]
```diff
@@ -36,13 +36,14 @@ symlinkJoin {
...
-    desktop=share/applications/org.fcitx.Fcitx5.desktop
-    autostart=etc/xdg/autostart/org.fcitx.Fcitx5.desktop
-    rm $out/$desktop
-    rm $out/$autostart
-    cp ${fcitx5}/$desktop $out/$desktop
-    sed -i $out/$desktop -e "s|^Exec=.*|Exec=$out/bin/fcitx5|g"
-    ln -s $out/$desktop $out/$autostart
+    pushd $out
+    grep_dirs=(share/applications etc/xdg/autostart share/dbus-1/services)
+    grep -Rl --include=\*.{desktop,service} "''${grep_dirs[@]}" -e ${fcitx5} | while read -r file; do
```
Comment (SuperSandro2000) [inline pkgs/tools/inputmethods/fcitx5/with-addons.nix:41]:
```suggestion
    grep -Rl --include=\*.{desktop,service} share/applications etc/xdg/autostart share/dbus-1/services -e ${fcitx5} | while read -r file; do
```
we can really easily avoid quoting hell here

Comment (duament) [reply pkgs/tools/inputmethods/fcitx5/with-addons.nix]: done

**Review by Vonfry — APPROVED** [review]

**Pull request merged**